server-spi-private/src/main/java/org/keycloak/userprofile/UserProfile.java

```diff
@@ -17,16 +17,72 @@
 
 package org.keycloak.userprofile;
 
+import java.util.function.BiConsumer;
+
+import org.keycloak.models.UserModel;
+
 /**
- * Abstraction, which allows to update the user in various contexts (Required action of already existing user, or first identity provider
- * login when user doesn't yet exists in Keycloak DB)
+ * <p>An interface providing as an entry point for managing users.
+ *
+ * <p>A {@code UserProfile} provides a manageable view for user information that also takes into account the context where it is being used.
+ * The context represents the different places in Keycloak where users are created, updated, or validated.
+ * Examples of contexts are: managing users through the Admin API, or through the Account API.
+ *
+ * <p>By taking the context into account, the state and behavior of {@link UserProfile} instances depend on the context they
+ * are associated with, where validating, updating, creating, or obtaining representations of users is based on the configuration
+ * and constraints associated with a context.
  *
+ * <p>A {@code UserProfile} instance can be obtained through the {@link UserProfileProvider}.
+ *
+ * @see UserProfileContext
+ * @see UserProfileProvider
  * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
  */
 public interface UserProfile {
 
-    String getId();
+    /**
+     * Validates the attributes associated with this instance.
+     *
+     * @throws ValidationException in case
+     */
+    void validate() throws ValidationException;
+
+    /**
+     * Creates a new {@link UserModel} based on the attributes associated with this instance.
+     *
+     * @throws ValidationException in case validation fails
+     *
+     * @return the {@link UserModel} instance created from this profile
+     */
+    UserModel create() throws ValidationException;
+
+    /**
+     * <p>Updates the {@link UserModel} associated with this instance. If no {@link UserModel} is associated with this instance, this operation has no effect.
+     *
+     * <p>Before updating the {@link UserModel}, this method first checks whether the {@link #validate()} method was previously
+     * invoked. If not, the validation step is performed prior to updating the model.
+     *
+     * @param removeAttributes if attributes should be removed from the {@link UserModel} if they are not among the attributes associated with this instance.
```

Comment:
'boolean removeAttributes' parameter should not be here as UserProfile itself should encapsulate logic/decision for the attribute update. Caller context should be passed here so internal impl can decide what to do. We can probably change this later once we start to implement processing based on the configuration

Reply:
I agree. The only reason why I have that there is because of how our REST API works and for backward compatibility with the old console.

Wr.t. to the Admin REST API, attributes are only removed if the attribute map was sent (either empty or not). In case it is null, we don't remove attributes at all. We can't make this decision without relying on this condition from the caller.